source/topic3.rst

```diff
@@ -417,21 +418,197 @@ Contact List Class
+    * Overloading
+    * We can even do something called :doc:`constructor chaining </topic3-chaining>`
+
+* Since arrays have a fixed size, our strategy here is to make an array with a sufficiently large size, but only use what we need
+* The first constructor will make use of some constant value set in the class to make the array
+* The second will take a capacity as a parameter and make the array that size
+
+* If we create a ``ContactList`` object
+    * ``ContactList contacts = new ContactList(5);``
+
+* We will have something like this created
+
+    .. image:: ../img/contacts.png
+       :width: 600 px
+       :align: center
+
 
 Adding Friends
 ^^^^^^^^^^^^^^
 
+**Python**
+
+.. code-block:: python
+    :linenos:
+
+    def add(self, first_name, last_name, email):
+        # Make the friend object
+        new_friend = Friend(first_name, last_name, email)
+
+        # Append friend to our friends list
+        # and update friend count
+        self._friends.append(new_friend)
+        self._friends_count += 1
+
+
+**Java**
+
+.. code-block:: java
+    :linenos:
+    :emphasize-lines: 16, 17, 18
+
+    /**
+     * Add a new friend to the friends array. Will create an instance
+     * of a Friend based on parameters. If our array runs out of space
+     * we will expand capacity to manage the situation.
+     *
+     * @param firstName     friends first name
+     * @param lastName      friends last name
+     * @param email         friends email address
+     */
+    public void add(String firstName, String lastName, String email) {
+        // Create the Friend object
+        Friend newFriend = new Friend(firstName, lastName, email);
+
+        // If we have run out of space in our array
+        // we need to deal with it
+        if (friendCount == friends.length) {
+            expandCapacity();
+        }
+        // Add friend to the next available spot
+        friends[friendCount] = newFriend;
+        friendCount++;
+    }
+
+    private void expandCapacity() {
+        // Make a new array of twice the size of the previous
+        Friend[] newFriends = new Friend[friends.length * 2];
+
+        // Copy over the contents of the friends list
+        // to the new bigger friends list
+        for(int i = 0; i < friends.length; ++i) {
+            newFriends[i] = friends[i];
+        }
+        // Have friends now reference the new friends
+        friends = newFriends;
+    }
+
+* You may notice that ``friendCount`` plays double duty here --- friends count and next available spot in the array
+
+* Since our array has a fixed size, we can't simply keep adding to it
+* Our solution is to ``expandCapacity``
+    1. Create a new array twice as big as the original
+    2. Copy over the contents of the original array to the new bigger array
+        * `You could also use this instead <https://docs.oracle.com/javase/7/docs/api/java/lang/System.html#arraycopy%28java.lang.Object,%20int,%20java.lang.Object,%20int,%20int%29>`_
```

Comment:
Java 7 reference, definitely refer to jdk11 (or at least 8). Also, it's typical to prefer [Arrays.copyOf](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Arrays.html#copyOf(T%5B%5D,int)) over arraycopy.

Reply:
done and done.